DirectProgramming/DPC++FPGA/include/streaming_cholesky.hpp

```diff
@@ -0,0 +1,254 @@
+                            // Only lower-left elements of L are
+                            // sent in row order, starting with row 0.
+          >
+struct StreamingCholesky {
+  void operator()() const {
+    // Functional assertions 
+    static_assert(rows >= 4,
+                  "Only matrices of size 4x4 and over are supported");
+    static_assert(pipe_size >= 1,
+                  "The pipe must be able to contain at least one element");
+
+    // Set the computation type to T or ac_complex<T> depending on the value
+    // of is_complex
+    using TT = std::conditional_t<is_complex, ac_complex<T>, T>;
+
+    constexpr int kColumns = rows;
+
+    // Number of lower-left elements in the L output matrix
+    constexpr int kLMatrixSize = kColumns * (kColumns + 1) / 2;
+
+    // Compute Cholesky decompositions as long as matrices are given as inputs
+    while (1) {
+      // Break memories up to store pipe_size elements per bank
+      constexpr short kBankwidth = pipe_size * sizeof(TT);
+      constexpr unsigned short kNumBanks = rows / pipe_size;
+
+      // When specifying numbanks for a memory, it must be a power of 2.
+      // Unused banks will be automatically optimized away.
+      constexpr short kNumBanksNextPow2 =
+          fpga_tools::Pow2(fpga_tools::CeilLog2(kNumBanks));
+
+      [[intel::numbanks(kNumBanksNextPow2)]]  // NO-FORMAT: Attribute
+      [[intel::bankwidth(kBankwidth)]]        // NO-FORMAT: Attribute
+      [[intel::private_copies(4)]]            // NO-FORMAT: Attribute
+      [[intel::max_replicates(1)]]            // NO-FORMAT: Attribute
+      TT a_load[rows][kColumns];
+
+      // Two copies of L to be able to load two complete rows per iteration
+      // Multiple private copies to be able to overlap multiple loop
+      // iterations
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
+      TT l_result_compute[rows][kColumns];
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
+      TT l_result_compute_copy[rows][kColumns];
+
+      [[intel::private_copies(2)]]  // NO-FORMAT: Attribute
+      TT l_result[kLMatrixSize];
+
+      // Copy a matrix from the pipe to a local memory
+      // Number of pipe reads of pipe_size required to read a full column
+      constexpr int kExtraIteration = ((rows % pipe_size) != 0) ? 1 : 0;
+      constexpr int kLoopIterPerColumn = (rows / pipe_size) + kExtraIteration;
+      // Number of pipe reads of pipe_size to read all the matrices
+      constexpr int kLoopIter = kLoopIterPerColumn * kColumns;
+      // Size in bits of the loop iterator over kLoopIter iterations
+      constexpr int kLoopIterBitSize =
+          fpga_tools::BitsForMaxValue<kLoopIter + 1>();
+
+      [[intel::initiation_interval(1)]]  // NO-FORMAT: Attribute
+      for (ac_int<kLoopIterBitSize, false> li = 0; li < kLoopIter; li++) {
+        fpga_tools::NTuple<TT, pipe_size> pipe_read = AIn::read();
+
+        int write_idx = li % kLoopIterPerColumn;
```

Comment:
Will this be an actual mod operator? Or Is `kLoopIterPerColumn` always a power of 2? If it can be a non-power of 2, we could remove this expensive mod operator.

Reply:
It depends on the size of the matrix column and the number of elements transferred per pipe operation.
Both are user defined, and are known at compile time.
I'm guessing you are suggesting to implement this modulo operator using a truth table computed at compile time?

Reply:
Truth table could work if you expect `kLoopIterPerColumn` to be relatively small. Or simply hoist `write_idx` out of the loop. Something like this:

```cpp
int write_idx = 0;
for (ac_int<kLoopIterBitSize, false> li = 0; li < kLoopIter; li++) {
    // ...
    write_idx++;
    if (write_idx == kLoopIterPerColumn) write_idx = 0;
    // ...
}
```

This could probably be made better with shannonization, but maybe the compiler takes care of it.
Your choice! If this is not on the critical path, maybe it is not worth it, I just get scared whenever I see '%'.

Reply:
I should go over the qrd, qri and cholesky kernels to make this change in all the appropriate places.
Let's do this in a separte bulk change